biogtr/models/global_tracking_transformer.py

```diff
@@ -90,15 +92,33 @@
             An N_T x N association matrix
         """
         # Extract feature representations with pre-trained encoder.
-        for frame in filter(
-            lambda f: f.has_instances() and not f.has_features(), frames
+        if any(
+            [
+                (not instance.has_features()) and instance.has_crop()
+                for instance in ref_instances
+            ]
         ):
-            crops = frame.get_crops()
-            z = self.visual_encoder(crops)
+            ref_crops = torch.concat(
+                [instance.crop for instance in ref_instances], axis=0
+            )
+            ref_z = self.visual_encoder(ref_crops)
+            for i, z_i in enumerate(ref_z):
+                ref_instances[i].features = z_i
 
-            for i, z_i in enumerate(z):
-                frame.instances[i].features = z_i
+        if query_instances:
+            if any(
+                [
+                    (not instance.has_features()) and instance.has_crop()
+                    for instance in query_instances
+                ]
+            ):
+                query_crops = torch.concat(
+                    [instance.crop for instance in query_instances], axis=0
+                )
+                query_z = self.visual_encoder(query_crops)
+                for i, z_i in enumerate(query_z):
+                    query_instances[i].features = z_i
```

Comment:
Do this! Could be useful to have it as an instance method so that you can call it by itself as well for different kinds of preprocessing pipelines.

It may also be helpful to have it not be all-or-nothing and instead just run on the ones without features (or have a flag in this function to recompute the features explicitly).

Reply:
@talmo if I make it a class method it won't have access to `self.visual_encoder`